manifest/store.go

```diff
@@ -0,0 +1,115 @@
+package manifest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/docker/pkg/ioutils"
+	"github.com/opencontainers/go-digest"
+)
+
+// Store provides methods which can operate on a manifest store.
+type Store interface {
+	Get(digest.Digest) (*distribution.Manifest, error)
+	Set(manifest distribution.Manifest) (digest.Digest, error)
+	Prune() error
+}
+
+// store implements a Store on the filesystem.
+type store struct {
+	sync.RWMutex
+	// root is the directory under which manifests are cached.
+	root string
+}
+
+const (
+	contentDirName = "content"
+)
+
+// NewManifestStore returns a new filesystem backed Store.
+func NewManifestStore(root string) (Store, error) {
+	abspath, err := filepath.Abs(root)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &store{root: abspath}
+	if err := os.MkdirAll(filepath.Join(root, contentDirName, string(digest.Canonical)), 0700); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *store) contentFile(dgst digest.Digest) string {
+	return filepath.Join(s.root, contentDirName, string(dgst.Algorithm()), dgst.Hex())
+}
+
+// Get returns the content stored under a given digest.
+// The content is verified to match the lookup digest.
+func (s *store) Get(dgst digest.Digest) (*distribution.Manifest, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	data, err := ioutil.ReadFile(s.contentFile(dgst))
+	if err != nil {
+		return nil, err
+	}
+
+	if digest.FromBytes(data) != dgst {
+		return nil, fmt.Errorf("failed to verify cached manifest: %v", dgst)
+	}
+
+	var version manifest.Versioned
+	if err := json.Unmarshal(data, &version); err != nil {
+		return nil, err
+	}
+
+	m, _, err := distribution.UnmarshalManifest(version.MediaType, data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// Set stores content by digest.
+func (s *store) Set(manifest distribution.Manifest) (digest.Digest, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	_, data, err := manifest.Payload()
+	if err != nil {
+		return "", err
+	}
+
+	dgst := digest.FromBytes(data)
+	if err := ioutils.AtomicWriteFile(s.contentFile(dgst), data, 0600); err != nil {
+		return "", err
+	}
+
+	return dgst, nil
+}
+
+// Prune clears all cached manifest data.
+func (s *store) Prune() error {
```

Comment:
@tonistiigi Good point. I just pushed a commit to modify the behavior. It's a little tricky since image deletion is so flexible.